### PR TITLE
add genmap/mappability

### PR DESCRIPTION
**Pull request by jianhong**
## PR checklist

- [ x ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ x ] If necessary, include test data in your PR.
- [ x ] Remove all TODO statements.
- [ x ] Emit the `<SOFTWARE>.version.txt` file.
- [ x ] Follow the naming conventions.
- [ x ] Follow the parameters requirements.
- [ x ] Follow the input/output options guidelines.
- [ x ] Add a resource `label`
- [ x ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ x ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ x ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: software/genmap/mappability/main.nf [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    } else {
+        container "quay.io/biocontainers/genmap:1.3.0--h1b792b2_1"
+    }
+
+    input:
+    path fasta
+
+    output:
+    path "*.wig"                  , emit: wig
+    path "*.version.txt"          , emit: version
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:26]:
```suggestion
    path "*.wig"        , emit: wig
    path "*.version.txt", emit: version
```

File: software/genmap/mappability/main.nf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:30]:
```suggestion
```

File: software/genmap/mappability/main.nf [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    genmap index -F $fasta -I genmap_index
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:32]: I think we should probably have a separate module for `genmap index`?

File: software/genmap/mappability/main.nf [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    genmap map \\
+        $options.args \\
+        -I genmap_index \\
+        -O . \\
+        -w
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:37]:
```suggestion
```

Hmmm...maybe we should have optional outputs depending on whether `-w`/ `-bg` / `-t` are provided as opposed to hard-coding wig files as output?
https://github.com/cpockrandt/genmap#computing-the-mappability

We can check which options have been passed to the module using logic like this:
https://github.com/nf-core/modules/blob/4575e5455c62560e0cabdf128ec3adcc81f84a2c/software/rsem/preparereference/main.nf#L33

and then have optional outputs for each of these like here:
https://github.com/nf-core/modules/blob/4575e5455c62560e0cabdf128ec3adcc81f84a2c/software/star/align/main.nf#L34

**Review by drpatelh — COMMENTED** [review]
Looks awesome @jianhong ! 🤩 A few comments to maybe re-organise the module logic and split out `genmap index` into a separate module.

File: software/genmap/index/main.nf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    input:
+    path fasta
+
+    output:
+    tuple path(fasta), path("genmap_index/index.*"), emit: index
+    path "*.version.txt"                           , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+
```
Comment (drpatelh) [inline software/genmap/index/main.nf:30]:
```suggestion
```

File: software/genmap/index/main.nf [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+params.options = [:]
+options        = initOptions(params.options)
+
+process GENMAP_INDEX {
+    tag '$fasta'
+    label 'process_high'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:[:], publish_by_meta:[]) }
```
Comment (drpatelh) [inline software/genmap/index/main.nf:12]:
```suggestion
        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:'index', meta:[:], publish_by_meta:[]) }
```

Comment (drpatelh) [reply software/genmap/index/main.nf:12]: For consistency with other indexing modules e.g. [bwa index](https://github.com/nf-core/modules/blob/master/software/bwa/index/main.nf)

File: software/genmap/index/main.nf [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    } else {
+        container "quay.io/biocontainers/genmap:1.3.0--h1b792b2_1"
+    }
+
+    input:
+    path fasta
+
+    output:
+    tuple path(fasta), path("genmap_index/index.*"), emit: index
+    path "*.version.txt"                           , emit: version
```
Comment (drpatelh) [inline software/genmap/index/main.nf:26]:
```suggestion
	path "genmap"       , emit: index
    path "*.version.txt", emit: version
```

Comment (drpatelh) [reply software/genmap/index/main.nf:26]: For consistency with other indexing modules e.g. [bwa index](https://github.com/nf-core/modules/blob/master/software/bwa/index/main.nf)

File: software/genmap/index/main.nf [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+
+    """
+    genmap index \\
+        -F $fasta \\
+        -I genmap_index
```
Comment (drpatelh) [inline software/genmap/index/main.nf:34]:
```suggestion
    genmap 
        index \\
        -F $fasta \\
        -I genmap
```

File: software/genmap/mappability/main.nf [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    conda (params.enable_conda ? "bioconda::genmap=1.3.0" : null)
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/genmap:1.3.0--h1b792b2_1"
+    } else {
+        container "quay.io/biocontainers/genmap:1.3.0--h1b792b2_1"
+    }
+
+    input:
+    tuple path(fasta), path(index)
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:22]:
```suggestion
    path index
```

File: software/genmap/mappability/main.nf [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+    output:
+    path "mappability.{wig,bedgraph,txt}", emit: map
+    path "*.version.txt"                 , emit: version
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:26]:
```suggestion
    path "*.wig"        , optional:true, emit: wig
    path "*.bedgraph"   , optional:true, emit: bedgraph
    path "*.txt"        , optional:true, emit: txt
    path "*.version.txt", emit: version
```

File: software/genmap/mappability/main.nf [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    """
+    genmap map \\
+        $options.args \\
+        -I . \\
+        -O mappability
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:34]:
```suggestion
    genmap \\
        map \\
        $options.args \\
        -I . \\
        -O mappability
```

File: software/genmap/mappability/meta.yml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+      description: index file
+
+output:
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - map:
+      type: file
+      description: mappability file
```
Comment (drpatelh) [inline software/genmap/mappability/meta.yml:30]:
```suggestion
  - wig:
      type: file
      description: genmap wig mappability file
	  pattern: "*.wig"
  - bedgraph:
      type: file
      description: genmap bedgraph mappability file
	  pattern: "*.bedgraph"
  - txt:
      type: file
      description: genmap text mappability file
	  pattern: "*.txt"
```

File: tests/software/genmap/mappability/main.nf [hunk 13]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { GENMAP_INDEX       } from '../../../../software/genmap/index/main.nf' addParams( options: [:] )
+include { GENMAP_MAPPABILITY } from '../../../../software/genmap/mappability/main.nf' addParams( options: [args : '-K 50 -E 2 -w -t -bg'] )
+
```
Comment (drpatelh) [inline tests/software/genmap/mappability/main.nf:7]:
```suggestion
include { GENMAP_INDEX       } from '../../../../software/genmap/index/main.nf'       addParams( options: [:] )
include { GENMAP_MAPPABILITY } from '../../../../software/genmap/mappability/main.nf' addParams( options: [args : '-K 50 -E 2 -w -t -bg'] )

```

File: software/genmap/mappability/main.nf [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    path "*.version.txt"               , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    """
+    genmap \\
+        map \\
+        $options.args \\
+        -I . \\
+        -O mappability
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:37]:
```suggestion
	INDEX=`find -L ./ -name "*.ids.concat" -printf '%h\n'`

    genmap \\
        map \\
        $options.args \\
        -I \$INDEX \\
        -O mappability
```

File: software/genmap/mappability/main.nf [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    script:
+    def software = getSoftwareName(task.process)
+    """
+    INDEX=`find -L ./ -name "*.ids.concat" -printf '%h\n'`
+
+    genmap \\
+        map \\
+        $options.args \\
+        -I \$INDEX \\
+        -O mappability
```
Comment (drpatelh) [inline software/genmap/mappability/main.nf:39]:
```suggestion
    genmap \\
        map \\
        $options.args \\
        -I $index \\
        -O mappability
```

**Review by drpatelh — APPROVED** [review]
Thanks @jianhong ! I changed a few bits to simplify and standardise the modules with respect to everything else we have on nf-core/modules.

**Comment by jianhong** [comment]
Thank you @drpatelh for all updates!

**Pull request merged**